### PR TITLE
dockerless nitro-cli

### PR DESCRIPTION
**Pull request by calincoman**
*Description of changes:*
- Added new functions for handling the cache logic
- Fixed issues raised by Petre and Erdem
- Fixed several cargo clippy errors (remaining errors are unused functions / constants)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: enclave_build/src/utils.rs [hunk 1]
```diff
@@ -106,15 +110,59 @@ impl ExtractLogic {
 
         // Try to parse the JSON
         let json_object: Value = serde_json::from_str(config_string.as_str()).unwrap();
-        let config_obj = json_object.get("container_config").expect("'container_config' field is missing");
-        let cmd_obj = config_obj.get("Cmd").expect("'Cmd' field is missing");
+        let config_obj = json_object.get("container_config").expect("'container_config' field is missing.");
+        let cmd_obj = config_obj.get("Cmd").expect("'Cmd' field is missing.");
```
Comment (petreeftime) [inline enclave_build/src/utils.rs:114]: Should return errors instead of crashing here.

File: enclave_build/src/image_cache.rs [hunk 2]
```diff
@@ -1,57 +1,265 @@
...
+        for expr in expressions {
+            let res = writeln!(&mut writer, "{}", expr);
+            match res {
+                Ok(_) => (),
+                Err(err) => {
+                    return Err(CacheError::DataCacheError(format!(
+                        "Failed to write {} expression to the output cache file: {}", expression_name, err)));
+                }
+            };
+        }
```
Comment (petreeftime) [inline enclave_build/src/image_cache.rs:250]: Replace with
```rust
expressions.map(|expr| writeln!(...).map_error(...)).collect()?;
```

File: enclave_build/src/image_cache.rs [hunk 3]
```diff
@@ -1,57 +1,265 @@
...
+        let output_path = vec![path.clone(), match expression_name.as_str() {
+            "ENV" => constants::ENV_CACHE_FILE_NAME.to_string(),
+            "CMD" => constants::CMD_CACHE_FILE_NAME.to_string(),
+            // This case was already handled above
+            _ => {
+                return Err(CacheError::ArgumentError("".to_string()));
+            },
+        }].concat();
+        // Create the file
+        let output_file = match File::create(output_path) {
```
Comment (petreeftime) [inline enclave_build/src/image_cache.rs:229]: use map_err

File: enclave_build/src/cache_manager.rs [hunk 4]
```diff
@@ -29,25 +38,49 @@ impl CacheManager {
...
+        &mut self.values
+    }
+
     /// Return the image hash corresponding to the image URI if available, or else return an error
-    pub fn get_image_hash(&mut self, uri: &String) -> Result<String, CacheError> {
-        let image_hash = match self.values.entry(uri.to_string()) {
-            Entry::Occupied(val) => Some(val.get().clone()),
-            Entry::Vacant(_) => None
+    pub fn get_image_hash(&self, uri: &String) -> Result<String, CacheError> {
+        let image_hash: Option<&String> = match self.values.get(uri) {
```
Comment (petreeftime) [inline enclave_build/src/cache_manager.rs]: I don't understand what this match does?

**Review by petreeftime — CHANGES_REQUESTED** [review]
Can you run `cargo clippy` over the code and fix the issues?

File: enclave_build/src/image_cache.rs [hunk 5]
```diff
@@ -310,7 +324,7 @@ impl CachePath {
 
     /// Returns the path to the cache root folder of an image
     /// e.g. {ROOT}/.nitro_cli/container_cache/{IMAGE_HASH}
-    pub fn get_image_path(image_ref: &Reference, cache_manager: &CacheManager) -> Result<String, CacheError> {
+    pub fn get_image_folder_path(image_ref: &Reference, cache_manager: &CacheManager) -> Result<String, CacheError> {
```
Comment (petreeftime) [inline enclave_build/src/image_cache.rs:327]: Should return `Path` or `PathBuf` instead of String.

File: enclave_build/src/image_cache.rs [hunk 6]
```diff
@@ -289,16 +290,29 @@ pub struct CachePath {}
...
+    /// 
+    /// For CACHE_ROOT_FOLDER = XDG_DATA_DIRS
+    // pub fn get_cache_root_folder() -> String {
+    //     let aux = env::var_os(CACHE_ROOT_FOLDER).unwrap();
+    //     let xdg_data_dirs = aux.to_str().unwrap();
+    //     let dirs: Vec<&str> = xdg_data_dirs.split(":").collect();
+
+    //     // For CACHE_ROOT_FOLDER = XDG_DATA_DIRS, choose to take the first directory path
+    //     // PROBLEM - on the EC2 AL2 instance XDG_DATA_DIRS is not set from what I have seen
+    //     // => should have some default root folder for this case
```
Comment (petreeftime) [inline enclave_build/src/image_cache.rs:303]: `$XDG_DATA_HOME defines the base directory relative to which user-specific data files should be stored. If $XDG_DATA_HOME is either not set or empty, a default equal to $HOME/.local/share should be used. `

https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

Comment (petreeftime) [reply enclave_build/src/image_cache.rs:303]: `$XDG_DATA_DIRS defines the preference-ordered set of base directories to search for data files in addition to the $XDG_DATA_HOME base directory. The directories in $XDG_DATA_DIRS should be seperated with a colon ':'.

If $XDG_DATA_DIRS is either not set or empty, a value equal to /usr/local/share/:/usr/share/ should be used. `

For writing, I would XDG_DATA_HOME or the default, if this is not set. When searching for a container image, the cli should also look into XDG_DATA_DIRS, but I would add this later on.

File: enclave_build/src/cache_manager.rs (hunk 4 above)
Comment (petreeftime) [reply enclave_build/src/cache_manager.rs]: This whole function can be re-written as
```rust
self.values.get(uri).map(|s| s.to_string()).ok_or_else(|err| ...)
```

**Comment by sabin-rapan** [comment]
Please write decent commit messages and titles.
Also fix the CI failures

**Pull request closed without merging**